### ISSUE TITLE
Enhancement: [no-unsafe-*] check `any`s nested inside object arguments

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#11801]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unsafe-argument

### Description

Currently `no-unsafe-argument` doesn't catch object arguments with nested `any` types. For example:

```ts
// Imagine this is a value received from a library.
declare const fromLib: { foo: any };

// Imagine this is a function defined in user-land.
declare const fn: (x: { foo: number }) => string;

// This is unsafe due to the inner `any` type.
fn(fromLib);
```

A real world example of where this comes up is React Router's `useLocation` hook, which returns a `Location` object where the `state` property has type `any`.

Would it make sense to extend this lint rule to check for `any`s nested inside object types?

### Fail

```TypeScript
// Imagine this is a value received from a library.
declare const fromLib: { foo: any };

// Imagine this is a function defined in user-land.
declare const fn: (x: { foo: number }) => string;

// This is unsafe due to the inner `any` type.
fn(fromLib);
```

### Pass

```TypeScript
declare const fromLib: { foo: number };

declare const fn: (x: { foo: number }) => string;

fn(fromLib);
```

**Comment by OliverJAsh** [comment]
Interestingly this works if you use a shared generic `interface`, but not if you use a `type` alias or inline type. [Playground](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgGQm0xgQiQB4AVAPjgG84BnGE4ALjirgF8AoPgHpBcAJIBbTAHNkeGAAsETRMsxwAbpgA2AVzxRguBOuAATOOigRxcNVoQAjKJigBPAHR9ThrS7zZyFgsrcXxHTkJiUnIKTCRXGgBuAWExSRkkOUVlJVsLHSRsaKQ4b3RZc2Q4HSY0AFpfJFNPb2xfAzgApCD0JE4AChAIohIySiQdcQc0GgBKOABeOhYoZClkoREqbJVq7sx0PFM9OBgIU-k8ZEyoOAADONc709cwYE9e-strMIdZxKAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9oAc1gBbRC3RRE0aB2iRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

**Comment by bradzacher** [comment]
The reason the generic interface works is because it's a very simple check.
For `T<U> <- V<W>` -- if `T == V` then you just check `U != any && V == any`

Fun fact -- this is similar to how TS actually does fast-path generic checks itself -- if the base type is the same then it skips all other checks and just does "is `V` assignable to `U`"!

This algorithm is trivial for us to implement and lets us support all manner of cases with little effort.

The issue with doing deep object checks is that instead of an `O(1)` top-level check per type like we've got right now, you instead have an `O(N)` check per type as you need to iterate through all of the properties looking for `any`s.

Not saying I'm against this -- there's definite value in the added safety!
We would just want to make sure we implement it carefully to avoid exploding the rule's cost.

We'd also not just want this for `no-unsafe-argument`, and instead for all `no-unsafe-*` rules.
